### PR TITLE
FINERACT-38

### PR DESCRIPTION
**Pull request by maektwain**
## Description
I was able to merge the source code which was done @Ippezrobert  on FINERACT-38 , I have done some minimal testing as of now. 

Functional testing needs to be done by @MayaDuruva

Have not written test as of now.

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [ Y] Commit message starts with the issue number from https://issues.apache.org/jira/projects/FINERACT/. Ex: FINERACT-646 Pockets API.

- [ Y ] Coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions have been followed.

- [ N] API documentation at https://github.com/apache/fineract/blob/develop/api-docs/apiLive.htm has been updated with details of any API changes.

- [ N] Integration tests have been created/updated for verifying the changes made.

- [ Y] All Integrations tests are passing with the new commits.

- [ N] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the list.)

Our guidelines for code reviews is at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide

### CODE REVIEW

**Comment by xurror** [comment]
This build breaks some checkstyle rules.
Can you check that out?

**Comment by maektwain** [comment]
@xurror Just finished the checkStyleMain task , it seems like the IDE by its own does not picks up the file defined in config , plus it did not fix the files , had to do manually, should be a task where we can fix the files like fixCheckStyleMain 

though my test are still failing locally, but as I see they are passing on travis.

**Review by xurror — APPROVED** [review]

**Comment by vorburger** [comment]
> Functional testing needs to be done by @MayaDuruva

@MayaDuruva can you confirm that you will be performing functional testing on this PR?

> Have not written test as of now.

@maektwain Including an integration test for this functionality (in this PR) would be great.

**Comment by maektwain** [comment]
This loos like pushed away due to COVID-19, I would like to finish the functional testing and one member is helping me finishing the integration test.

**Comment by maektwain** [comment]
@vorburger hey , I was able to get some of the tests contributed by one of the GSOC aspirant, maybe down the day we would test the functionality. Could you help speed this up so that we can merge it quickly.

**Comment by vorburger** [comment]
@maektwain this PR now has merge conflicts which need to be manually resolve with a rebase...

**Comment by maektwain** [comment]
Hello @vorburger I did a quick rebase from develop branch to FINERACT-38 , let me know if this solves the conflict.

**Comment by vorburger** [comment]
> Hello @vorburger I did a quick rebase from develop branch to FINERACT-38 , let me know if this solves the conflict.

@maektwain nope it does not... note the "This branch cannot be rebased due to conflicts" message, shown by GitHub just above? Also notice the build failure.. following #750 (based on #740) you have to remove that generics in your new class.

You'll probably want the build to fully to pass locally first after a rebase before you push, and double check yourself that GitHub does not show you "This branch cannot be rebased due to conflicts" anymore.

**Comment by maektwain** [comment]
@vorburger Nopes , it does not show me any conflict issues , quite strange. About the generics I have removed them.
Screenshot 2020-04-15 at 4 11 28 PM

**Comment by maektwain** [comment]
Oops, I know what is the issue, I took me while to realise the I was rebasing my own repo so actually this was happening 
Apache Fineract -- My Repo -- all the while I was pulling only My Repo.

**Comment by maektwain** [comment]
I think, there are many commits now. This will solve all of the issue now.

**Comment by maektwain** [comment]
Have done functional testing also I was able to move and close the accounting closure. Seems to be working fine. 

There is a problem though with SQL Group by query which failed in 5.7 , I fixed it by setting it off , though this is not the bets route to do it.

```
Expression #1 of SELECT list is not in GROUP BY clause and contains nonaggregated column
```

**Comment by maektwain** [comment]
ok, I have fixed the above issue.

File: fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/ClientSavingsIntegrationTest.java [hunk 1]
```diff
@@ -1907,7 +1908,7 @@ public void testSavingsAccount_WITH_WITHHOLD_TAX_DISABLE_AT_ACCOUNT_LEVEL() {
         Assert.assertEquals(expected, actual, 1);
 
     }
-
+    @Ignore
```
Comment (vorburger) [inline fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/ClientSavingsIntegrationTest.java]: remove this from this PR, and rebase from master, you'll get it from #761 ...

File: fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/CalculateIncomeAndExpanseBookingIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+    @Before
+    public void setup() {
+        calculateIncomeAndExpenseBooking =  new CalculateIncomeAndExpenseBookingImpl(null, null, null, null, incomeAndExpenseReadPlatformService,officeReadPlatformService);
+     }
+
+    @After
+    public void tearDown() {
+
+    }
```
Comment (vorburger) [inline fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/CalculateIncomeAndExpanseBookingIntegrationTest.java]: remove this empty method:

```suggestion
```

File: fineract-provider/src/main/java/org/apache/fineract/organisation/office/service/OfficeReadPlatformServiceImpl.java [hunk 3]
```diff
@@ -266,4 +266,20 @@ public OfficeTransactionData retrieveNewOfficeTransactionDetails() {
...
+        sqlBuilder.append(" JOIN m_office ounder ON ounder.hierarchy LIKE CONCAT(mo.hierarchy, '%') ");
+        sqlBuilder.append(" AND ounder.hierarchy like CONCAT('.', '%') where mo.id = ? ");
+
+
+        try {
+            return this.jdbcTemplate.queryForList(sqlBuilder.toString(), Long.class,
+                    new Object[] { officeId});
+        } catch (final EmptyResultDataAccessException e) {
+           return null;
+        }
```
Comment (vorburger) [inline fineract-provider/src/main/java/org/apache/fineract/organisation/office/service/OfficeReadPlatformServiceImpl.java]: This strikes me as curious - is this done like this elsewhere in existing code?

**Review by vorburger — CHANGES_REQUESTED** [review]
Had a quick first review, and have 3 minor feedback. Also there was a test failure in `StorageTest` which is most unlikely unrelated to this PR, but a known flaky test, see https://issues.apache.org/jira/browse/FINERACT-828 ... could I suggest that you rebase and we see if it is green?

File: fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/CalculateIncomeAndExpanseBookingIntegrationTest.java (hunk 2 above)
Comment (maektwain) [reply fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/CalculateIncomeAndExpanseBookingIntegrationTest.java]: @singkara could you fix this issue.

File: fineract-provider/src/main/java/org/apache/fineract/organisation/office/service/OfficeReadPlatformServiceImpl.java (hunk 3 above)
Comment (maektwain) [reply fineract-provider/src/main/java/org/apache/fineract/organisation/office/service/OfficeReadPlatformServiceImpl.java]: @singkara

File: fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/ClientSavingsIntegrationTest.java (hunk 1 above)
Comment (maektwain) [reply fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/ClientSavingsIntegrationTest.java]: @vorburger ack done.

**Comment by maektwain** [comment]
```
org.apache.fineract.notification.StorageTest > testNotificationStorage FAILED
    org.mockito.exceptions.verification.opentest4j.ArgumentsAreDifferent at StorageTest.java:125
```
Is this supposed to fail

**Comment by vorburger** [comment]
> StorageTest

@maektwain see https://issues.apache.org/jira/browse/FINERACT-828 about that one (search JIRA!), do you want to see if you can help to fix that one, in a separate PR? The (latest build of this) PR doesn't fail because of that test though, but due to https://issues.apache.org/jira/browse/FINERACT-899... do you know how we could fix that?

Perhaps more importantly than those test failure, this PR has merge conflicts now, and a weird Commit history... 44 commits, of which many which are already on `develop` now - I think something went wrong during your `git rebase`.. do you want to update this to have 1 clean commit only?

**Comment by maektwain** [comment]
Hi @singkara , So this has become really big, if you have time maybe you can do this, 

rebase all the things on the latest develop branch , in just a single commit. Could you do this?

**Comment by maektwain** [comment]
@vorburger Flaky tests could mean some depending values and arguments are changing, perhaps might need to investigate.

Do we have more information about it?

**Comment by vorburger** [comment]
@maektwain all known information is in JIRA, in the tickets linked to https://issues.apache.org/jira/browse/FINERACT-850 - there is one for each known flaky test. Exactly WHY each of those are flaky is what working on those JIRAs requires to find out... :smile_cat:

**Comment by maektwain** [comment]
@singkara could you do this , we close this PR and we might need to just add this commit.

7f9967a
142da4f
2fa66ce
19783b1
091f186

That is all, and add your own tests also and replay it on the our repo before we do open a PR here, and we try to do a squash in single commit.

Let me know if you have any question.

**Comment by singkara** [comment]
Hi, @maektwain I have created the new branch FINERACT-38_rebase for the new changes. However, while cherrypicking, I had merge conflicts in multiple files and since I am not the writer, I am not sure which ones to keep. Perhaps we could get on a call and finish this.
The attached screenshot shows the conflicts just because of cherrypicking 1st commit.
Screen Shot 2020-04-19 at 19 37 33
Although I have an intuition of which changes shall go in, just to be cautious I would ask you to review the merge conflicts once.

**Comment by maektwain** [comment]
Take the head one, and try to replay and lets close this off.

**Comment by vorburger** [comment]
@maektwain @singkara see (new) https://github.com/apache/fineract#pull-requests and do not wait for an active maintainer to come and help you re. build failures - it's your responsibility to get a green build. Have you checked the log, and done what the README suggests to do in such cases? :smiley_cat:

**Pull request closed without merging**